internal/route/websocket.go

```diff
@@ -0,0 +1,113 @@
+package route
+
+import (
+	"log"
+	"net/http"
+
+	"github.com/gorilla/websocket"
+)
+
+// Message that will be sent to client
+type Message struct {
+	Event string `json:"event"`
+	Data  struct {
+		ListID string `json:"list_id"`
+		TodoID string `json:"todo_id"`
+		Done   bool   `json:"done"`
+	} `json:"data"`
+}
+
+// RegisterMessage registers client
+type RegisterMessage struct {
+	Message Message
+	Client  *websocket.Conn
+}
+
+type socketManager struct {
+	clients   map[string][]*websocket.Conn
+	register  chan RegisterMessage
+	broadcast chan Message
+}
+
+var manager = socketManager{
+	clients:   make(map[string][]*websocket.Conn),
+	register:  make(chan RegisterMessage), // registration channel
+	broadcast: make(chan Message),         // broadcast channel
+}
+
+var upgrader = websocket.Upgrader{}
+
+// HandleWs handles websocket connections
+func HandleWs(w http.ResponseWriter, r *http.Request) {
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer ws.Close()
+
+	for {
+		var msg Message
+		err := ws.ReadJSON(&msg)
+
+		if err != nil {
+			log.Printf("error: %v\n", err)
+			break
+		}
+
+		if msg.Event == "register" {
+			registerMsg := RegisterMessage{
+				Message: msg,
+				Client:  ws,
+			}
+			manager.register <- registerMsg
+		} else {
+			manager.broadcast <- msg
+		}
+	}
+}
+
+// HandleMessages processes WS messages and sends them to the client
+func HandleMessages() {
+	for {
+		select {
+		case msg := <-manager.broadcast:
+			manager.handleMessage(&msg)
+		case reg := <-manager.register:
+			manager.registerClient(&reg)
```

Comment:
Since this is a func that's intended to run in a Goroutine, it's suggested to use a method name that implies that. Something like `ListenForWebsocketMessages` or `StartWebsocketHandler` etc...

Also, given that it's your top-level func. Think of a "sub-main()", it's a "main" for its Goroutine, so it should probably log.

have `manager.handleMessage(...)` and `manager.registerClient(...)` return errs and `log.Printf` them, bubble everything up to here.

Reply:
Another thing that might make error bubbling click is logging. When you deploy into production you a `fmt.Println` or `log.Printf` isn't that useful. Sure you can read the log of the container, but nowadays we tend to use something that ships the logs off.

So say we had a `newrelic.Logger` we'd wire it up in our app's `main()` then we'd need to send it to all the interested parties that could log.

So with that you can consider error bubbling. If an error controls flow and needs to ship the message to New Relic, the idea of sending it down to a model struct is kinda ugly. You want to send it to as few places as possible.

With that we _could_ send it to `StartWebsocketHandler`, that's probably OK. Make sense?